File: cpp/src/vnd.cpp
#include <vector>
#include <memory>
#include <functional>
#include "neighborhoods.hpp"
#include "solvers.hpp"
#include "utils.hpp"

Solution VND::vnd(
    const Instance& I,
    const Solution& initial_sol,
    const Neighborhood::NeighborhoodFactory &neighborhood_factories
) {
    Solution sol = initial_sol;
    double f = utils::objective(I, sol);

    int i = 0;
    int K = (int)neighborhood_factories.size();

    while (i < K) {

        Neighborhood::NeighborhoodFactory single_neigh = { neighborhood_factories[i] };
        MaxIterations stopping(100000);
        Solution new_sol = LS::local_search(
            I,
            sol,
            single_neigh,
            StepFunction::first_improvement,
            stopping
        );

        double f_new = utils::objective(I, new_sol);

        if (f_new < f) {
            sol = new_sol;
            f = f_new;
            i = 0;
        } else {
            i += 1;
        }
    }

    return sol;
}

File: python/vnd.py
from neighborhoods import *
from local_search import (
    local_search,
    StepFunction,
    StoppingCriterion,
    ImprovementThreshold,
    MaxIterations,
)


def vnd(I: Instance, sol: Solution, neighborhoods_cls: List[type[Neighborhood]]):
    sol = deepcopy(sol)
    f = objective(I, sol)

    i = 0
    K = len(neighborhoods_cls)

    while i < K:
        new_sol = local_search(
            I,
            sol,
            [neighborhoods_cls[i]],
            StepFunction.first_improvement,
            MaxIterations(10**5),
        )

        f_new = objective(I, new_sol)

        if f_new < f:
            sol = new_sol
            f = f_new
            i = 0
        else:
            i += 1

    return sol
